openlibrary/catalog/add_book/__init__.py

```diff
@@ -678,6 +685,197 @@ def load_data(rec, account_key=None):
     return reply
 
 
+def update_matched_edition(
+    e: Edition, rec: dict, need_edition_save: bool, account_key: str | None = None
+) -> tuple[Edition, bool]:
+    """
+    Update `e` fields to add values from `rec` that don't currently
+    exist on `edition`. The default is to prefer existing values rather than
+    to overwrite.
+    """
+    # Add cover to `e`, and remove `cover` from `rec`, because it's not
+    # a valid `Edition` field and would otherwise be copied in.
+
+    cover_url = rec.pop('cover', None)
+    if cover_url and not e.get_covers():
+        if cover_id := add_cover(cover_url, e.key, account_key=account_key):
+            e['covers'] = [cover_id]
+            need_edition_save = True
+
+    # Convert the record to an OpenLibrary Edition suitable for saving, so
+    # that values copied from the record into the Edition are of the proper
+    # format.
+    rec_edition = build_query(rec)
+
+    # Update `e` str and int value fields with `rec` values. Don't overwrite
+    # existing fields by default.
+    for field, value in rec_edition.items():
+        if (
+            not (
+                isinstance(value, str)
+                or (
+                    isinstance(value, dict)
+                    and value.get("type") == ('/type/text' or '/type/int')
+                )
+            )
+            or field in e
+        ):
+            continue
+
+        e[field] = value
+        need_edition_save = True
+
+    # Promise items sometimes get "????" as `publish_date`; replace if
+    # something else is available.
+    if e.publish_date == "????" and (
```

Comment:
We should never write `????` as a `date` value in the first place.
It might be better to resolve this by cleaning up any existing `????` dates and updating import code to reject them in future.

Reply:
I've removed this and I will update this PR tomorrow to reflect that, after I obsessively go over each line again only to miss something obvious. :)

I will also try to figure out how `????` has been sneaking into `publish_date`.

Reply:
I opened a PR (#7519) for the `publish_date="????"` issue.